### DISCUSSION TITLE
Proper way to search for term1 AND term2

### DISCUSSION THREAD

**Discussion by hiqua** [BurntSushi/ripgrep#1845] [Q&A] [answered]
Hi,
from what I've read, (rip)grep does not seem to have an AND connector, and I've seen suggestions such as `grep -E 'pattern1.*pattern2|pattern2.*pattern1' filename`. That involves typing each pattern twice. My lazy option right now is `rg term1 | rg term2` which doesn't play well with `rg -l`.

The best solution I can imagine right now is defining a shell function taking pattern1 pattern2 as inputs and make it run `grep -E 'pattern1.*pattern2|pattern2.*pattern1' filename`.

Is there a better way? Would it be an appropriate feature for rg?

**Comment by BurntSushi** [comment, marked as answer]
> from what I've read, (rip)grep does not seem to have an AND connector

That's correct. There is an open issue tracking it: https://github.com/BurntSushi/ripgrep/issues/875

> and I've seen suggestions such as `grep -E 'pattern1.*pattern2|pattern2.*pattern1' filename`

Yup, that is indeed a work-around. Works "fine" for two patterns, but does not scale.

> My lazy option right now is `rg term1 | rg term2` which doesn't play well with `rg -l`.

Yes. My general hesitance with adding boolean things like `--and` is that `rg term1 | rg term2` generally capture it pretty well _and_ it scales. The main problem with that is that most tools these days (including ripgrep) have totally abandoned the UNIX philosophy, and thus, piping `rg term1 | rg term2` doesn't quite work for all use cases. `-l` is certainly one of them, although, one can work-around that with a bit more typing:

```
rg term1 -l -0 | xargs -0 rg term2 -l
```

Of course, this prints files that contain `term1` and `term2` anywhere, rather than files with at least one line that contains both `term1` and `term2`. So it's not quite the same unfortunately.

> The best solution I can imagine right now is defining a shell function taking pattern1 pattern2 as inputs and make it run `grep -E 'pattern1.*pattern2|pattern2.*pattern1' filename`.

Yup, pretty much. Or use another tool that supports boolean combinations like `git grep`. (While I haven't done a thorough investigation myself, my guess is that ripgrep's boolean matching interface will be inspired by `git grep`'s.)

**Reply by hiqua** [reply]
Thanks a lot for the thorough answer! I'll follow the issue, and use the alias in the meantime.

**Comment by learnbyexample** [comment]
You can use `-P` option here, for example:

```bash
# match pat1 and pat2
rg -P '(?=.*pat1).*pat2'

# match pat1 and pat2 and pat3
rg -P '(?=.*pat1)(?=.*pat2).*pat3'

# match pat1 and pat2 but NOT rejectpat1
rg -P '^(?!.*rejectpat1)(?=.*pat1).*pat2'
```

**Reply by hiqua** [reply]
Maybe I'm missing something, but it seems you're talking about the OR connector, while I'm looking for AND.

**Reply by learnbyexample** [reply]
Oops, my comments used 'or' instead of 'and'. I've edited it now. The solutions are for AND only.

**Reply by hiqua** [reply]
Apparently the Debian build doesn't include PCRE2 support so I can't test your solutions, but for instance I don't expect:
```
rg -P '(?=.*pat1).*pat2'
```
to match `pat2   pat1`, does it?

If not we're back to what has been mentioned in the OP and in the previous answer (which indicates that its' indeed not supported right now so I don't think you'll be able to find something significantly better than what I mentioned before).

**Reply by learnbyexample** [reply]
It will match. `(?=regexp)` is a positive lookahead assertion which checks if that pattern matched without consuming characters. You can test these patterns with `perl` or GNU grep as shown below:

```bash
$ s='pat2 pat1\npat1 pat2\npat1 pat3\n'

# note that third line doesn't match since pat2 wasn't found
$ printf '%b' "$s" | rg -P '(?=.*pat1).*pat2'
pat2 pat1
pat1 pat2

$ printf '%b' "$s" | grep -P '(?=.*pat1).*pat2'
pat2 pat1
pat1 pat2

$ printf '%b' "$s" | perl -ne 'print if /(?=.*pat1).*pat2/'
pat2 pat1
pat1 pat2
```

**Reply by hiqua** [reply]
Nice, thanks!
